common/clock/ratelimiter.go

```diff
@@ -0,0 +1,381 @@
+// or the times we compare are not guaranteed to be monotonic due to waiting on mutexes.
+// The `.Before` comparison should ensure *better* behavior in spite of this, but it's
+// much more correct if it's acquired within the lock.
+func (r *ratelimiter) lockNow() (now time.Time, unlock func()) {
+	r.mut.Lock()
+	newNow := r.timesource.Now()
+	if r.latestNow.Before(newNow) {
+		// this should always be true because `time.Now()` is monotonic, and it is
+		// always acquired inside the lock (so values are strictly ordered and
+		// not arbitrarily delayed).
+		//
+		// that means e.g. lockCorrectNow should not ever receive a value newer than
+		// the Now we just acquired, so r.latestNow should never be newer either.
+		//
+		// that said: it still shouldn't be allowed to rewind.
+		r.latestNow = newNow
+	}
+	return r.latestNow, r.mut.Unlock
+}
+
+// lockCorrectNow gets the correct "now" time to use, and ensures it never rolls back.
+// this is intended to be used by reservations, to allow them to return claimed tokens
+// after arbitrary waits, as long as no other calls have interleaved.
+//
+// the returned value MUST be used instead of the passed time: it may be the same or newer.
+// the passed value may rewind the rate.Limiter's internal time record, and cause undefined behavior.
+//
+// The lock MUST be held until all time-accepting methods are called on the
+// underlying rate.Limiter, as otherwise the internal "now" value may rewind
+// and cause undefined behavior.
+func (r *ratelimiter) lockCorrectNow(tryNow time.Time) (now time.Time, unlock func()) {
+	r.mut.Lock()
+	if r.latestNow.Before(tryNow) {
+		// this should be true if a cancellation arrives before any other call
+		// has advanced time (e.g. other Allow() calls).
+		//
+		// "old" cancels are expected to skip this path, and return the r.latestNow
+		// without updating it.  they may or may not release their token in this case.
+		r.latestNow = tryNow
+	}
+	return r.latestNow, r.mut.Unlock
+}
+
+func (r *ratelimiter) Allow() bool {
+	now, unlock := r.lockNow()
+	defer unlock()
+	return r.limiter.AllowN(now, 1)
+}
+
+func (r *ratelimiter) Burst() int {
+	return r.limiter.Burst()
+}
+
+func (r *ratelimiter) Limit() rate.Limit {
+	return r.limiter.Limit()
+}
+
+func (r *ratelimiter) Reserve() Reservation {
+	now, unlock := r.lockNow()
+	defer unlock()
+	res := r.limiter.ReserveN(now, 1)
+	if res.OK() && res.DelayFrom(now) == 0 {
+		// usable token, return a cancel-able handler in case it's not used
+		return &reservation{
+			res:        res,
+			reservedAt: now,
+			limiter:    r,
+		}
+	}
+	// unusable, immediately cancel it to get rid of the future-reservation
+	// since we don't allow waiting for it.
+	//
+	// this restores MANY more tokens when heavily contended, as time skews
+	// less before canceling, so it gets MUCH closer to the intended limit,
+	// and it performs very slightly better.
+	res.CancelAt(now)
+	return deniedReservation{}
```

Comment:
noticed some super-low allow rates in a mock-time benchmark, tracked the issue down to "cancels often arrive later than desired in this test", leading to the ratelimiter mostly staying at negative tokens and only allowing ~0.001% through.

synchronously canceling solves that, and should also improve behavior IRL.